_posts/2024-01-22-Python-Memory-Management-and-Troubleshooting.md

```diff
@@ -0,0 +1,542 @@
+---
+title: "Tip of the Month: Python Memory Management and Troubleshooting"
+author: dave-bunten
+tags:
+  - tip-of-the-month
+  - python
+  - memory
+  - memory-management
+  - memory-allocators
+---
+
+# Tip of the Week: Python Memory Management and Troubleshooting
+
+{% include tip-of-the-month-intro.html %}
+
+## Introduction
+
+<!-- excerpt start -->
+Have you ever run Python code only to find it taking _forever_ to complete or sometime abruptly ending with an error like: `123456 Killed` or `killed (program exited with code: 137)`?
+You may have experienced memory resource or management challenges associated with these scenarios.
+This post will cover some computer memory definitions, how Python makes use of computer memory, and share some tools which may help with these types of challenges.
+<!-- excerpt end -->
+
+## What is Memory?
+
+### Computer Memory
+
+```mermaid!
+flowchart LR
+
+subgraph computer ["Computer (resources)"]
+  memory["Memory"]
+  storage["Data Storage"]
+  cpu["CPU(s)"]
+end
+
+style computer fill:#fff,stroke:#333
+style storage fill:#fff,stroke:#333
+style cpu fill:#fff,stroke:#333
+style memory fill:#86EFAC,stroke:#333
+```
+
+_Computer memory is a type of computer resource available for use by software on a computer_
+
+Computer memory, also sometimes known as "RAM" or "random-access memory", or "dynamic memory" is a type of resource used by computer software on a computer.
+"Computer memory stores information, such as data and programs for immediate use in the computer. ... Main memory operates at a high speed compared to non-memory storage which is slower but less expensive and oftentimes higher in capacity. " ([Wikipedia: Computer memory](https://en.wikipedia.org/wiki/Computer_memory)).
+
+<table>
+<tr><th colspan="2">Memory Blocks</th></tr>
+<tr>
+<td>
+
+<strong>A.)</strong> All memory blocks available.
+
+<table>
+<tr><td>Block</td><td>Block</td><td>Block</td></tr>
+</table>
+
+</td>
+<td>
+
+<strong>B.)</strong> Some memory blocks in use.
+
+<table>
+<tr><td style="background:#86EFAC;">Block</td><td style="background:#86EFAC;">Block</td><td>Block</td></tr>
+</table>
+
+</td>
+</tr>
+<tr>
+<td colspan="2" style="text-align:center;font-weight:bold;">Practical analogy</td>
+</tr>
+<tr>
+<td>
+
+<strong>C.)</strong> You have limited buckets to hold things.
+
+<table>
+<tr><td>🪣</td><td>🪣</td><td>🪣</td></tr>
+</table>
+
+</td>
+<td>
+
+<strong>D.)</strong> Two buckets are used, the other remains empty.
+
+<table>
+<tr><td style="background:#86EFAC;">🪣</td><td style="background:#86EFAC;">🪣</td><td>🪣</td></tr>
+</table>
+
+</td>
+</tr>
+</table>
+
+_Fixed-size memory blocks may be free or used at various times. They can be thought of like reusable buckets to hold things._
+
+One way to organize computer memory is through the use of ["fixed-size blocks"](https://en.wikipedia.org/wiki/Memory_management#FIXED-SIZE), also called "blocks".
```

Comment:
Thanks, great comment! I'm not certain of how the CPU interacts at this level, this absolutely needs a follow up with more of a deep dive at the CPU level.

On word before diagram: with this post generally my approach is to include a visual aid and describe it using text below. I find this makes for a good way to address many different kinds of readers - those who may find the visual helpful to contextualize the text, and those who may read the text first while comparing it with what the visual contains. Hopefully a "two-pass" / multi-modal approach for many, either way, to increase comprehension 🙂 . Personally I find this can sometimes help entice a reader to keep going with heavier material in the form of a "visual break" in the text to keep the cognitive load somewhat paced (I don't have strong scientific evidence of this however).

Reply:
That wikipedia link to fixed-size blocks mentions its application in embedded systems, which Python pretty much isn't.  Looking at the Python page for [Memory Mangement](https://docs.python.org/3/c-api/memory.html#c.PyMemAllocatorDomain), especially the first code snippet, I see they're using the word block to refer to what you get back from malloc. Note that you can pass whatever size you want in for BUFSIZ, not a pre-specified size.  It's just a chunk of memory that you will think of as something more abstract seconds later. In the example, the result of the malloc() call is cast to "char *", a character pointer, basically a string.  Another allocation might cast the returned block/chuck/amount of memory as a "struct Person", int, float, double or whatever. From then on it's not just a block (of memory) so much as a Person structure, integer, floating point number etc.

The block returned by malloc, is however, fixed in the sense that it, until you realloc() it, is the same size it was when you create it. See also [buffer overruns](https://en.wikipedia.org/wiki/Buffer_overflow).

Reply:
....and it may need to be said that a Person struct, int, float and String are basically different sizes.